### PR TITLE
Integration of executors in Functor Filter

### PR DESCRIPTION
**Pull request by shrijitsingh99**
TODO:
Improve the merging of indices (https://github.com/shrijitsingh99/pcl/pull/10#discussion_r460382209)

### CODE REVIEW

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
If no more update is needed, please break this into 2 commits too
* filter indices + filter base classes
* all filters

**Comment by shrijitsingh99** [comment]
> break this into 2 commits

Rebased with the base branch `executor-integration`, so now the commit structure should be correct.

File: filters/include/pcl/filters/functor_filter.h [hunk 1]
```diff
@@ -16,17 +16,18 @@ namespace pcl {
 template <typename PointT, typename Function>
 constexpr static bool is_functor_for_filter_v =
     pcl::is_invocable_r_v<bool,
-                          Function,
-                          const pcl::remove_cvref_t<pcl::PointCloud<PointT>>&,
-                          pcl::index_t>;
+        Function,
+        const pcl::remove_cvref_t<pcl::PointCloud<PointT>>&,
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]: Remove the `remove_cvref_t`. It's not needed since you have the type right there

File: filters/include/pcl/filters/functor_filter.h [hunk 2]
```diff
@@ -76,22 +77,48 @@ class FunctorFilter : public FilterIndices<PointT> {
    */
   void
   applyFilter(Indices& indices) override
+  {
+    const auto exec = executor::best_fit{};
+    applyFilter(exec, indices);
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]:
```suggestion
    applyFilter(std::move(exec), indices);
```
Since we don't need this anymore

File: filters/include/pcl/filters/functor_filter.h [hunk 3]
```diff
@@ -76,22 +77,48 @@ class FunctorFilter : public FilterIndices<PointT> {
...
-      // functor returns true for points that should be selected
-      if (negative_ != functor_(*input_, index)) {
-        indices.push_back(index);
-      }
-      else if (extract_removed_indices_) {
-        removed_indices_->push_back(index);
+    auto cond_filtering = [&](executor::executor_index_t<Executor> index) {
+      pcl::utils::ignore(index);
+      #pragma omp for
+      for (index_t idx = 0; idx < indices_->size(); ++idx) {
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]: OMP wants signed index. Please don't use `index_t` here

File: filters/include/pcl/filters/functor_filter.h [hunk 4]
```diff
@@ -76,22 +77,48 @@ class FunctorFilter : public FilterIndices<PointT> {
...
+        exec.execute([&]() {
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]: Why is executor executing this? This is the job of "internal executor" which has been on low priority IIRC

File: filters/include/pcl/filters/functor_filter.h [hunk 5]
```diff
@@ -76,22 +77,48 @@ class FunctorFilter : public FilterIndices<PointT> {
...
+        });
       }
+    };
+
+    exec.bulk_execute(cond_filtering, 0);
+    merge_indices(exec, indices, keep);
+  }
+
+  template <typename Executor, typename executor::instance_of_base<Executor, executor::inline_executor, executor::omp_executor> = 0>
+  void merge_indices(const Executor &exec, Indices& indices, std::vector<std::uint8_t>& keep) {
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]: Maybe a better name (and does it need to be public)?

File: test/filters/test_functor_filter.cpp [hunk 6]
```diff
@@ -94,6 +94,59 @@ INSTANTIATE_TEST_SUITE_P(RandomSeed,
                          FunctorFilterRandom,
                          testing::Values(123, 456, 789));
 
+
+template <typename T>
```
Comment (kunaltyagi) [inline test/filters/test_functor_filter.cpp]: This test is almost a copy paste. We need to think how we should test executor using code

**Review by kunaltyagi — COMMENTED** [review]
Commit 1

File: filters/include/pcl/filters/functor_filter.h [hunk 7]
```diff
@@ -9,6 +9,9 @@
 
 #pragma once
 
+#define PCL_NO_PRECOMPILE
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/functor_filter.h]: Errr.... What's this code supposed to do? The include is alive in L15

Moreover, not including the header will cause a compile error, right?

File: filters/include/pcl/filters/functor_filter.h [hunk 8]
```diff
@@ -9,24 +9,28 @@
...
+#include <pcl/filters/filter_indices.h>
+#undef PCL_NO_PRECOMPILE
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/functor_filter.h]: You're unconditionally defining PCL_NO_PRECOMPILE and then undefining it. If a user defines it in his/her code and then includes this filter, we're gonna mess up his/her definitions. 

tl;dr: remember the user can define PCL_NO_PRECOMPILE before you

File: filters/include/pcl/filters/functor_filter.h [hunk 9]
```diff
@@ -9,24 +9,28 @@
...
 /**
  * \brief Filter point clouds and indices based on a functor passed in the ctor
  * \ingroup filters
  */
 template <typename PointT, typename Functor>
-class FunctorFilter : public FilterIndices<PointT> {
-  using Base = FilterIndices<PointT>;
+class FunctorFilter : public FilterIndices<PointT, FunctorFilter<PointT, Functor>> {
+  using Self = FunctorFilter<PointT, Functor>;
+  using Base = FilterIndices<PointT, FunctorFilter<PointT, Functor>>;
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/functor_filter.h]: Use `Self`

File: filters/include/pcl/filters/functor_filter.h (hunk 2 above)
Comment (SergioRAgostinho) [reply filters/include/pcl/filters/functor_filter.h]: Or go directly for
```suggestion
    applyFilter(executor::best_fit{}, indices);
```

File: filters/include/pcl/filters/functor_filter.h (hunk 3 above)
Comment (SergioRAgostinho) [reply filters/include/pcl/filters/functor_filter.h]: Do we need to define an equivalent `sindex_t`, analogous to `uindex_t`?

File: filters/include/pcl/filters/functor_filter.h [hunk 10]
```diff
@@ -76,22 +80,48 @@ class FunctorFilter : public FilterIndices<PointT> {
...
+        });
       }
+    };
+
+    exec.bulk_execute(cond_filtering, 0);
+    merge_indices(exec, indices, keep);
+  }
+
+  template <typename Executor, typename executor::instance_of_base<Executor, executor::inline_executor, executor::omp_executor> = 0>
+  void merge_indices(const Executor &exec, Indices& indices, std::vector<std::uint8_t>& keep) {
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/functor_filter.h]: out parameters is a code-smell in modern C++. Construct and return `indices` by value inside this method. Trust RVO.
`keep` can be defined const.
definitely not a public method.

File: test/filters/test_functor_filter.cpp (hunk 6 above)
Comment (SergioRAgostinho) [reply test/filters/test_functor_filter.cpp]: Indeed, might as well try to extend the original test and reuse as much code as possible. 

besides being a carbon copy of the original test, you're testing with a single seed and you need need to always validate behavior with 2 or more seeds.

File: filters/include/pcl/filters/functor_filter.h (hunk 10 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/functor_filter.h]: That doesn't work for variables that might be holding a resource (like memory). out-parameters can reuse the allocated memory but new variable can't.

See: https://godbolt.org/z/s4Yf5d

File: filters/include/pcl/filters/functor_filter.h (hunk 3 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/functor_filter.h]: Seems likely. I was using `std::ptrdiff_t` but `sindex_t` appears better appraoch

File: filters/include/pcl/filters/functor_filter.h (hunk 8 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Fixed with 98e44f7

File: filters/include/pcl/filters/functor_filter.h (hunk 7 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Fixed with 98e44f7 as discussed

File: filters/include/pcl/filters/functor_filter.h (hunk 5 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Made private. Would `extractMarkedIndices` be better?

File: filters/include/pcl/filters/functor_filter.h (hunk 4 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Was there in the PoC. Don't need it as of now since internal executors don't exist as of now.

File: filters/include/pcl/filters/functor_filter.h (hunk 3 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Switched to `std::ptrdiff_t` as of now. We can switch to `sindex_t` in a separate PR, I reckon there will be other places needing this update too,

File: filters/include/pcl/filters/functor_filter.h (hunk 2 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: On hold since with the new best fit in v0.3 of the executor will not require this. Will update the method after v0.3 is integrated.

File: filters/include/pcl/filters/functor_filter.h (hunk 10 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: > variables that might be holding a resource (like memory)

We don't know how much space the user has reserved before hand.

File: test/filters/test_functor_filter.cpp (hunk 6 above)
Comment (shrijitsingh99) [reply test/filters/test_functor_filter.cpp]: Combined the test 99c86ee.
Not the most elegant way but get the job done. It tests using multiple seeds all the executors as well as checks best fit.

**Comment by shrijitsingh99** [comment]
I believe all the requested changes have been made in this PR. Now the only blocker is the changes needed from executor v0.3

File: filters/include/pcl/filters/functor_filter.h (hunk 2 above)
Comment (shrijitsingh99) [reply filters/include/pcl/filters/functor_filter.h]: Integrate new API

File: filters/include/pcl/filters/experimental/functor_filter.h [hunk 11]
```diff
@@ -77,22 +75,50 @@ class FunctorFilter : public FilterIndices<PointT> {
...
-      if (negative_ != functor_(*input_, index)) {
-        indices.push_back(index);
-      }
-      else if (extract_removed_indices_) {
-        removed_indices_->push_back(index);
+    auto cond_filtering = [&](executor::executor_index_t<Executor> index) {
+      pcl::utils::ignore(index);
+#pragma omp for
+      for (std::ptrdiff_t idx = 0; idx < indices_->size(); ++idx) {
+        if (negative_ != functor_(*input_, idx)) {
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/experimental/functor_filter.h]: there's a bug here
option A
```suggestion
      for (std::size_t idx = 0; idx < indices_->size(); ++idx) {
        if (negative_ != functor_(*input_, (*indices)[idx])) {
```

option B
```suggestion
      for (uindex_t idx = 0; idx < indices_->size(); ++idx) {
        if (negative_ != functor_(*input_, (*indices)[idx])) {
```

File: filters/include/pcl/filters/experimental/functor_filter.h [hunk 12]
```diff
@@ -77,22 +75,50 @@ class FunctorFilter : public FilterIndices<PointT> {
...
       }
+    };
+
+    exec.bulk_execute(cond_filtering, 0);
+    extractMarkedIndices(exec, indices, keep);
+  }
+
+private:
+  template <typename Executor, typename executor::InstanceOfAny<Executor, executor::inline_executor, executor::omp_executor> = 0>
+  void extractMarkedIndices(const Executor &exec, Indices& indices, const std::vector<std::uint8_t>& keep) {
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/experimental/functor_filter.h]: Nothing here makes use of the executor. Remove all the unnecessary stuff.

File: filters/include/pcl/filters/experimental/functor_filter.h [hunk 13]
```diff
@@ -77,22 +75,50 @@ class FunctorFilter : public FilterIndices<PointT> {
...
+    pcl::utils::ignore(exec);
+    for (index_t i = 0; i < keep.size(); ++i) {
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/experimental/functor_filter.h]:
```suggestion
    for (uindex_t i = 0; i < keep.size(); ++i)
```

or
```suggestion
    for (std::size_t i = 0; i < keep.size(); ++i)
```

File: test/filters/test_functor_filter.cpp [hunk 14]
```diff
@@ -35,13 +35,20 @@ TEST(FunctorFilterTrait, CheckCompatibility)
   EXPECT_TRUE((is_functor_for_filter_v<PointXYZ, decltype(const_ref_all)>));
 }
 
-struct FunctorFilterRandom : public testing::TestWithParam<std::uint32_t> {
+template <typename T>
+class FunctorFilterExecutor : public ::testing::Test {
+public:
```
Comment (SergioRAgostinho) [inline test/filters/test_functor_filter.cpp]: just leave it as a struct like it originally was

File: test/filters/test_functor_filter.cpp [hunk 15]
```diff
@@ -35,13 +35,20 @@ TEST(FunctorFilterTrait, CheckCompatibility)
...
   void
   SetUp() override
+  {
+    cloud = make_shared<PointCloud<PointXYZ>>();
+  }
+
+  void
+  fillCloud(std::uint32_t seed)
   {
     cloud = make_shared<PointCloud<PointXYZ>>();
```
Comment (SergioRAgostinho) [inline test/filters/test_functor_filter.cpp]: replace for `cloud->clear()` here.

**Review by SergioRAgostinho — CHANGES_REQUESTED** [review]

File: filters/include/pcl/filters/experimental/functor_filter.h (hunk 11 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/experimental/functor_filter.h]: We need to add a test for a filter with missing indices to catch such errors
